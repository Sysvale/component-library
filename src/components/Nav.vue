<template>
	 <b-nav
	 	class="px-2"
		:class="oldSchool ? 'bg-cs' : 'bg-liga'"
	 >
		<component
			v-for="(item, i) in computedItems"
			:is="dropdownOrSingleItem(item)"
			:class="getClass(item)"
			:id="getElementKey(item, i)"
			:key="getElementKey(item, i)"
			:text="item.label"
			:active="isActive(item)"
			class="mr-4 mt-2"
			menu-class="py-1"
			@click="handleClick(item)"
		>
			<template
				v-if="isDropdown(item)"
			>
				<b-dropdown-item
					v-for="(subitem, j) in item.items"
					:id="getElementKey(subitem, j, true)"
					:key="getElementKey(subitem, j, true)"
					:active="isActive(subitem)"
					link-class="py-2"
					class="py-1"
					@click.stop="handleClick(subitem, item)"
				>
					{{ subitem.label }}
				</b-dropdown-item>
			</template>
			<template
				v-else
			>
				{{ item.label }}
			</template>
		</component>
	</b-nav>
</template>

<script>
export default {
	props: {
		items: {
			type: Array,
			default: () => [],
			description: 'A list of the menu itens to be displayed',
			required: true,
		},
		oldSchool: {
			type: Boolean,
			default: true,
			description: 'Defines the style of the menu. The new one is light themed and the old one follows the style used by the "Cidade SaudÃ¡vel"',
		},
		activeItem:  {
			type: Object,
			default: () => {},
			description: 'The currently active menu iten',
			required: true,
		},
	},

	data() {
		return {
			internalActiveItem: this.activeItem,
		}
	},

	methods: {
		handleClick(item) {
			this.internalActiveItem = item;
			this.$emit('click', this.internalActiveItem);
		},

		isActive(item) {
			return Object.keys(this.internalActiveItem).length > 0 ? this.internalActiveItem.path === item.path : false;
		},

		getElementKey(item, index) {
			return `${item.label.replace(/\s/g, '')}${index}`;
		}
	}
};
</script>
<style>
	a {
		outline: none;
	}

	.cs-mode .nav-link:hover {
		color: rgb(255, 255, 255);
	}

	.cs-mode .active {
		border-bottom: 4px solid rgb(58, 223, 124);
		background: rgb(90, 90, 90);
		border-radius: 4px  4px 0px 0px;
	}

	.bg-liga {
		background: linear-gradient(0.93deg, #FFFFFF 16.45%, #F1F5FB 105.67%);
	}

	.bg-cs {
		background: rgb(68, 68, 68);
	}

	.cs-mode .nav-link.active{
		color: rgb(255, 255, 255);		
	}

	.cs-mode .nav-link {
		color: rgb(208, 210, 213);
	}

	.liga-mode .active {
		border-bottom: 4px solid rgb(40, 90, 185);
		background: rgba(40, 90, 185, 0.05);
		border-radius: 4px  4px 0px 0px;
	}

	.liga-mode .nav-link.active{
		color: rgb(20, 32, 50);
	}

	.liga-mode .nav-link {
		color: rgb(106, 117, 128);
	}

	.liga-mode .nav-link:hover {
		color: rgb(20, 32, 50);
	}
</style>
