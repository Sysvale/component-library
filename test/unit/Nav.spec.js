import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import Nav from '../../src/components/Nav.vue';
import lodash from 'lodash';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

const mockedData = [
	{ label: 'Dummy label 1', path: 'dummy/path/1' },
	{ label: 'Dummy label 2', path: 'dummy/path/2' },
	{ label: 'Dummy label 3', path: 'dummy/path/3' },
	{
		label: 'Dummy option with subitems',
		items: [
			{
				label: 'Dummy item 1',
				path: '/dummy/sub/item/1',
			},
			{
				label: 'Dummy item 2',
				path: '/dummy/sub/item/2',
			},
			{
				label: 'Dummy item 3',
				path: '/dummy/sub/item/3',
			},
		]
	}
];

test('Component is mounted properly', () => {
	const wrapper = mount(Nav, {
		localVue,
		propsData: {
			items: mockedData,
			activeItem: mockedData[0],
		},
	});
	expect(wrapper).toMatchSnapshot();
});

describe("Items styles test", () => {
	test('if has one item set as active', () => {
		const wrapper = mount(Nav, {
			localVue,
			propsData: {
				items: mockedData,
				activeItem: mockedData[0],
			},
		});
		expect(wrapper.findAll('.active').length).toBe(1);
	});

	test('if has one subitem set as active', () => {
		window._ = lodash;
		const wrapper = mount(Nav, {
			localVue,
			propsData: {
				items: mockedData,
				activeItem: mockedData[3].items[0],
			},
		});
		expect(wrapper.findAll('.active').length).toBe(1);
	});

	test('if the dropdown parent is set as active', () => {
		window._ = lodash;
		const wrapper = mount(Nav, {
			localVue,
			propsData: {
				items: mockedData,
				activeItem: mockedData[3].items[0],
			},
		});
		expect(wrapper.findAll('.active-parent').length).toBe(1);
	});

	test('if oldschool mode is setted properly', () => {
		const wrapper = mount(Nav, {
			localVue,
			propsData: {
				items: mockedData,
				oldSchool: true,
				activeItem: mockedData[0],
			},
		});

		expect(wrapper.findAll('.bg-cs').length).toBe(1);
		expect(wrapper.findAll('.cs-mode').length).toBe(mockedData.length);
	});
});

describe("Change active item event tests", () => {
	test('if a event is emited when the item is clicked', () => {

		const wrapper = mount(Nav, {
			localVue,
			propsData: {
				items: mockedData,
				activeItem: mockedData[0],
			},
		});

		const elementIndex = 1;

		wrapper
			.find(`#${wrapper.vm.getElementKey(mockedData[elementIndex], elementIndex)} a`)
			.trigger('click');

		expect(wrapper.emitted().click).toBeTruthy();
		expect(wrapper.emitted().click).toEqual([
			[
				mockedData[elementIndex],
			]
		]);
	});

	test('if a event is emited when the subitem is clicked', () => {
		window._ = lodash;
		const wrapper = mount(Nav, {
			localVue,
			propsData: {
				items: mockedData,
				activeItem: mockedData[0],
			},
		});

		wrapper
			.find(`#${wrapper.vm.getElementKey(mockedData[3].items[0], 0, true)}`)
			.trigger('click');

		expect(wrapper.emitted().click).toBeTruthy();
		expect(wrapper.emitted().click).toEqual([
			[
				mockedData[3].items[0],
			]
		]);
	});
});
