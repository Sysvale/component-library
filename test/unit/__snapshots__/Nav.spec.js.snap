// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<ul class="nav px-2 bg-cs">
  <li id="Dummylabel10" text="Dummy label 1" menu-class="py-1" class="nav-item mr-4 mt-2 cs-mode"><a href="#" target="_self" class="nav-link active">
      Dummy label 1
    </a></li>
  <li id="Dummylabel21" text="Dummy label 2" menu-class="py-1" class="nav-item mr-4 mt-2 cs-mode"><a href="#" target="_self" class="nav-link">
      Dummy label 2
    </a></li>
  <li id="Dummylabel32" text="Dummy label 3" menu-class="py-1" class="nav-item mr-4 mt-2 cs-mode"><a href="#" target="_self" class="nav-link">
      Dummy label 3
    </a></li>
  <li id="Dummyoptionwithsubitems3" class="nav-item b-nav-dropdown dropdown mr-4 mt-2 cs-mode" to="[object Object]"><a id="Dummyoptionwithsubitems3__BV_toggle_" role="button" aria-haspopup="true" aria-expanded="false" href="#Dummyoptionwithsubitems3" target="_self" class="nav-link dropdown-toggle"><span>Dummy option with subitems</span></a>
    <ul tabindex="-1" aria-labelledby="Dummyoptionwithsubitems3__BV_toggle_" class="dropdown-menu py-1">
      <li role="presentation" class="py-1"><a id="Dummyitem10-subitem" role="menuitem" href="#" target="_self" class="dropdown-item py-2">
          Dummy item 1
        </a></li>
      <li role="presentation" class="py-1"><a id="Dummyitem21-subitem" role="menuitem" href="#" target="_self" class="dropdown-item py-2">
          Dummy item 2
        </a></li>
      <li role="presentation" class="py-1"><a id="Dummyitem32-subitem" role="menuitem" href="#" target="_self" class="dropdown-item py-2">
          Dummy item 3
        </a></li>
    </ul>
  </li>
</ul>
`;
